cranelift/fuzzgen/src/function_generator.rs

```diff
@@ -0,0 +1,269 @@
+        Ok(CallConv::SystemV)
+    }
+
+    fn generate_type(&mut self) -> Result<Type> {
+        // TODO: It would be nice if we could get these directly from cranelift
+        let scalars = [
+            // IFLAGS, FFLAGS,
+            // B1, B8, B16, B32, B64, B128,
+            I8, I16, I32, I64,
+            // I128,
+            // F32, F64,
+            // R32, R64,
+        ];
+        // TODO: vector types
+
+        let ty = self.u.choose(&scalars[..])?;
+        Ok(*ty)
+    }
+
+    fn generate_abi_param(&mut self) -> Result<AbiParam> {
+        // TODO: Generate more advanced abi params (structs/purposes/extensions/etc...)
+        let ty = self.generate_type()?;
+        Ok(AbiParam::new(ty))
+    }
+
+    fn generate_signature(&mut self) -> Result<Signature> {
+        let callconv = self.generate_callconv()?;
+        let mut sig = Signature::new(callconv);
+
+        // TODO: Unconstrain this
+        for _ in 0..self.u.int_in_range(0..=8)? {
+            sig.params.push(self.generate_abi_param()?);
+        }
+
+        for _ in 0..self.u.int_in_range(0..=8)? {
+            sig.returns.push(self.generate_abi_param()?);
+        }
+
+        Ok(sig)
+    }
+
+    /// Creates a new var
+    fn create_var(&mut self, builder: &mut FunctionBuilder, ty: Type) -> Result<Variable> {
+        let id = self.vars.len();
+        let var = Variable::new(id);
+        builder.declare_var(var, ty);
+        self.vars.push((ty, var));
+        Ok(var)
+    }
+
+    fn vars_of_type(&self, ty: Type) -> Vec<Variable> {
+        self.vars
+            .iter()
+            .filter(|(var_ty, _)| *var_ty == ty)
+            .map(|(_, v)| *v)
+            .collect()
+    }
+
+    /// Get a variable of type `ty`, either reusing a old var, or generating a new one with a
+    /// `iconst`/`fconst`.
+    fn get_variable_of_type(
+        &mut self,
+        builder: &mut FunctionBuilder,
+        ty: Type,
+    ) -> Result<Variable> {
+        // TODO: global vars
+
+        let mut opts: Vec<
+            fn(
+                fgen: &mut FunctionGenerator,
+                builder: &mut FunctionBuilder,
+                ty: Type,
+            ) -> Result<Variable>,
+        > = Vec::with_capacity(2);
+
+        // Generate new
+        opts.push(|fgen, builder, ty| {
+            let imm64 = match ty {
+                I8 => fgen.u.arbitrary::<i8>()? as i64,
+                I16 => fgen.u.arbitrary::<i16>()? as i64,
+                I32 => fgen.u.arbitrary::<i32>()? as i64,
+                I64 => fgen.u.arbitrary::<i64>()?,
+                _ => unreachable!(),
+            };
+            let var = fgen.create_var(builder, ty)?;
+            let val = builder.ins().iconst(ty, imm64);
+            builder.def_var(var, val);
+
+            Ok(var)
+        });
+
+        // Reuse var
+        if self.vars_of_type(ty).len() != 0 {
+            opts.push(|fg, _, ty| {
+                let opts = fg.vars_of_type(ty);
```

Comment:
> when branching ... we choose a random number of variables ... to transfer onto the next block

This definitely works for a straightline sequence of blocks, but how does this work for loops? More generally, if you decide at the end of a block which variables to see as "defined" in the next block, you need to see all in-edges to a block before you know for sure that a variable is defined on every in-edge.

This is the origin of the two-stage approach -- it breaks the cycle: if you pin down what *should* be defined in each location, you can then, in parallel, both rely on that (in uses) and ensure that (in defs).

I think the easiest answer is to rely on the SSA-builder algorithm built into the variable helper layer which you're using here, and def everything at the top first, then select *any* variable for each use. The key thing (which I think may be flying under the radar here given the "transfer to the next block" phrasing above) is that "variables" as you're using them are not SSA values or blockparams; the latter are defined by the former depending on where you place the one or more definitions of a variable. (And indeed, a variable can have more than one def.) It's always correct to use a variable that has been defined at least once on every path from the entry block (has some dominating def), so the trivially correct thing to do is define everything (for the first time) in the entry block :-)

Reply:
I've updated the fuzzer to perform this upfront allocation of variables. I'm slightly concerned that it now seems to be reusing the same variable a lot consecutively, but I think this may be the fuzzer exploring the non ssa to ssa code?

Reply:
I hadn't seen your reply when writing my previous comment. I've changed it to the upfront allocation and i think it should be better because we are probably going to stress the variable helper a bit more with the fuzzer.

But i'm still not sure i'm following what you are saying. 

> This definitely works for a straightline sequence of blocks, but how does this work for loops? More generally, if you decide at the end of a block which variables to see as "defined" in the next block, you need to see all in-edges to a block before you know for sure that a variable is defined on every in-edge.

The way I'm thinking about this is as if every block is sort of a function.

```
function %test(i32, i32) -> i32 {
block0(v0: i32, v1: i32):
    v2 = iconst.i32 10
    jump block1(v1)  ; Here we can choose any of v0, v1, v2

block1(v4: i32):
    v5 = iconst.i32 12
    return v4              ; Here we can choose either v4 or v5 but never v0,v1,v2
}
```

So in a sense, what I was thinking was: We generate the next block (or reuse an existing one) see what types it has in its signature and choose variables available in the current block to "call" it with (either block params or values locally defined by us).

This way, we don't need to know who called this block, because we never use variables defined by anyone else.

This is more of a curiosity now, because I've changed into the other approach and think it is better, but I'd like to understand why it wouldn't work.

Reply:
> We generate the next block (or reuse an existing one) see what types it has in its signature and choose variables available in the current block to "call" it with (either block params or values locally defined by us).

Ah, I see! The key thing here, I think, is that "what types it has in its signature" is decided before code is generated, and so breaks the cycle: this is the equivalent to the pre-pass I mentioned that decides where to define things so that defs and uses can both rely on those decisions. So this approach would also work.

Given that the `cranelift-frontend` logic is built to accept multi-def vars and generate SSA, I think that the approach this PR takes now (upfront initializations) is still the right one, but your approach would work well for generating the SSA directly, if we decide to do that instead at some point.